[maahir-garg] iP

maahir-garg opened the pull request:
# The BotFather

> "Your mind is for having ideas, not holding them." – David Allen 

**BotFather** frees your mind of having to remember things you need to do. It's:

- Text-based
- Easy to learn
- **FAST, SUPER FAST** to use

All you need to do is:
1. download it from [here](https://github.com/maahir-garg/ip/releases/download/Bug-Fix/ip.jar).
2. run it from your command line using `java -jar ./ip.main.jar`
3. add your tasks.
4. let it manage your tasks for you 😉
*And it is FREE!*

## Features
- [x] Managing tasks
- [x] Managing deadlines 
- [ ] Reminders *(coming soon)*

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public static void main(String[] args) {
    new TheBotFather("./data/TheBotFather.txt").run();
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Level-0: Rename, Greet, Exit

Commit:
Level-1: Echo

Commit:
Level-2: Add, List

Commit:
Level-3: Mark as Done

Commit:
Level-4: ToDo, Event, Deadline

Commit:
UI Testing and more personality

Commit:
Level-5: Handle Errors

Commit:
Level-6: Delete

Commit:
Add Save functionality

Commit:
Merge branch 'branch-Level-7'

Commit:
Add DateTime functionality, Level-8

Commit:
Merge branch-Level-8 for DateTime functionality

Commit:
Create better OOP structuring

Commit:
Merge branch-OOP to add more OOp and improve structuring

Commit:
Create packages to better divide the classes

Commit:
Add gradle support

Commit:
Use gradle support

Commit:
Add J-Unit test support

Commit:
Create executable jar

Commit:
Add JavaDocs

Commit:
Improve styling

Commit:
Add JavaDocs

Commit:
Resolve merge conflicts in branch-A-CodingStandard

Commit:
Add Find option

Commit:
Resolve merge conflicts in branch-Level-9

Commit:
Add JavaDocs for Find and improve coding standards

TY1Fan commented on src/main/java/thebotfather/command/FindCommand.java:
```diff
@@ -0,0 +1,56 @@
+import thebotfather.util.TheBotFatherException;
+import thebotfather.util.Ui;
+
+import java.util.StringTokenizer;
```
I think the order of imports here is incorrect. java.util should be on top of thebotfather.util.

TY1Fan commented on src/main/java/thebotfather/TheBotFather.java:
```diff
@@ -0,0 +1,82 @@
+
+import thebotfather.command.Command;
+import thebotfather.task.Task;
+import thebotfather.util.*;
```
I think wildcard imports should be avoided.

TY1Fan commented on src/main/java/thebotfather/task/Task.java:
```diff
@@ -0,0 +1,118 @@
+    /**
+     * Formatter for printing date and time in a more readable format.
+     */
+    protected static final DateTimeFormatter DATE_STRING_FORMATTER_PRINT = DateTimeFormatter.ofPattern("dd MMM yyyy, HH:mm");
```
I think this line exceeds the hard [limit](https://se-education.org/guides/conventions/java/intermediate.html#layout) of 120 characters

TY1Fan commented on src/main/java/thebotfather/task/Task.java:
```diff
@@ -0,0 +1,118 @@
+    /**
+     * Indicates whether the task is completed.
+     */
+    private boolean done = false;
```
Would it be better to name this as isDone instead?

TY1Fan commented on src/main/java/thebotfather/util/Storage.java:
```diff
@@ -0,0 +1,102 @@
+                String line = fileReader.nextLine();
+                String[] instructions = line.split(" \\| ");
+                switch (instructions[0]) {
+                    case "T":
```
I think there should be no indentation for case clauses.

TY1Fan reviewed (commented):
I think overall not much issue. 

But need to take note of some nitty gritty details like:
- **Hard limit** of 120 characters per line 
- Import statement **orders**
- Avoid the use of _wildcard_ imports
- No indentation for case clauses
- And naming of booleans

Commit:
Update code according to new feedback

maahir-garg committed:
Fix indexing bug in mark, unmark and delete

benjiBase reviewed (commented):
LGTM! well done

maahir-garg committed:
Add CheckStyle

maahir-garg committed:
Add GUI

maahir-garg committed:
Add GUI

maahir-garg committed:
Add Assertions

The code lacks assertions to validate inputs and states, increasing the risk of runtime errors.

Add assertions to constructors and key methods to ensure valid inputs and states.

This ensures early detection of issues during development while keeping the code efficient in production.

maahir-garg committed:
Update CodeQuality

Document `DeadlineParser`, `EventParser`, `ExitParser`, `FindCommandParser`, `MarkTaskParser`, `TaskListParser`, `TodoParser`, and `UnmarkTaskParser` with explanations of their functionality, parameters, and exceptions.

Improve documentation for better clarity and maintenance.

maahir-garg committed:
Merge pull request #1 from maahir-garg/branch-A-Assertions

Add Assertions

maahir-garg committed:
Merge branch 'master' into branch-A-CodeQuality

maahir-garg committed:
Update TheBotFather.txt

maahir-garg committed:
Merge pull request #2 from maahir-garg/branch-A-CodeQuality

Branch a code quality

maahir-garg committed:
Add C-DetectDuplicates

maahir-garg committed:
Merge Extension to master

maahir-garg committed:
Add CI

maahir-garg committed:
Add CI

maahir-garg committed:
Update CI

maahir-garg committed:
Update CI

maahir-garg committed:
Update CI

maahir-garg committed:
Create README.md

maahir-garg committed:
Add title to GUI
